### DISCUSSION TITLE
Why not make `-o pipefail` the default for all shell scripts?

### DISCUSSION THREAD

**Discussion by tfenne** [nextflow-io/nextflow#3792] [General]
I recently got bitten by the fact that NF generates `.command.sh` scripts that _do_ set `-ue` but do not set `-o pipefail`.  Looking at the github issues and discussions, I see that other users are running into this too.

The fact that NF defaults to running `/bin/bash -ue` catches a lot of errors that arise in scripts and gives the user a sense that NF is catching _all_ execution failures and flagging processes as failed.

Being bioinformatics focused, I would expect to see piping frequently in NF workflows; this is certainly the case in workflows I write and review.  And also being bioinformatics focused, we are often running poorly-engineered tools that fail in weird and unexpected ways.

So my question is, why not make `/bin/bash -euo pipefail` the default in NF?  Is there anyone who would argue the case that they'd like pipe chains to be able to fail and the script still succeed (by default)?

**Comment by pditommaso** [comment]
Because it could introduce subtle bugs in existing pipelines. You can make pipefail the default by adding this line in your config 

```
process.shell = ['/bin/bash', '-euo', 'pipefail']
```

**Reply by tfenne** [reply]
Thanks @pditommaso - I understand I can do that (and have), but I think a lot of users coming to NF are going to get bitten by this.  Is there no process for making backward incompatible changes in NF?

**Reply by pditommaso** [reply]
Why they should `pipefail` is not the default behaviour in Bash. there are tricky situations also with it. See here 

https://github.com/nextflow-io/nextflow/issues/151

**Reply by tfenne** [reply]
I read through that issue, but unfortunately the gitter links with more context don't seem to work for me.  But that thread seems to go:

- Why not make it the default?
- There are issues ... it's not the default in bash
- Well, `set -e` is also not the default
- Ah, yes, but it might break existing pipelines

And I would echo that `set -e` and `set -u` are not default bash behavior, but are enabled by default in NF.  I understand the backwards compatibility argument - hence my question of whether there was a process for doing things like this in some major version rather than leaving a foot-gun around forever.

**Reply by pditommaso** [reply]
We'll consider it in a future major release
